adr/0002-masking-configuration-API.md

```diff
@@ -0,0 +1,199 @@
+
+The API for configuring `json-masker` will change to make it more consistent and extendable and allow customizations
+per JSON type.
+
+We will provide configuration options to customize all the defaults as well as the ability to restore the original
+behaviour of pre `1.0.0`.
+
+## Consequences
+
+Therefore the API will be changed to allow these masking options:
+```java
+public class MaskingConfig {
+    // Number masking, mutually exclusive options
+    private Boolean disableNumberMasking = false;    // disables number masking
+    private String maskNumbersWithString = "###";    // default
+    private Integer maskNumberWith;                  // mask number with another number, i.e. 0. Preserves the type 
+    private Integer maskNumberDigitsWith;            // mask all digits of a number with masking digit, must be 1 <= x <= 9 to avoid leading zeroes, preserves original length
+
+    // String masking, mutually exclusive options
+    private String maskStringsWith = "***";         // default
+    private Char maskStringCharactersWith;          // mask all characters with masking character, preserves original length
+
+    // Boolean masking, mutually exclusive options
+    private Boolean disableBooleanMasking = false;  // disables boolean masking
+    private String maskBooleansWithString = "&&&";  // default
+    private Boolean maskBooleanWith;                // mask boolean with another boolean, i.e. false. Preserves the type
+    
+    public static class Builder {
+        ...
+        public Builder disableNumberMasking() { ... };
+        public Builder maskNumbersWith(String) { ... };
+        public Builder maskNumbersWith(int) { ... };
+        public Builder maskNumberDigitsWith(int) { ... };
+
+        public Builder maskStringsWith(String) { ... };
+        public Builder maskStringCharactersWith(char) { ... };
+
+        public Builder disableBooleanMasking() { ... };
+        public Builder maskBooleanWith(String) { ... };
+        public Builder maskBooleanWith(boolean) { ... };
+    }
+} 
+```
+
+Current [JsonMaskingConfig](../src/main/java/dev/blaauwendraad/masker/json/config/JsonMaskingConfig.java) will be changed 
+to have a global per type configuration as well as per-key configuration:
+```java
+public class JsonmaskingConfig {
+    private final Set<String> targetKeys;
+    private final TargetKeyMode targetKeyMode;
+    private final MaskingConfig defaultConfig;
+    private final Map<String, MaskingConfig> targetKeyConfigs;
+    
+    public static class Builder {
+        ...
+
+        // mask mode
+        public Builder mask(String... targets) { ... };
+        public Builder mask(String target, MaskingConfig config) { ... };
+        public Builder mask(Set<String> targets, MaskingConfig config) { ... };
+        // hypothetical custom masker, most likely will be its own type instead of Function
+        public Builder mask(String target, Function<String | Number | Boolean, String> mapper) { ... };
```

Comment:
Such custom strategies will be nice to have. The type of the function is a bit tricky point though. If we are going to apply the function to the found key, then the type should be `byte[] -> byte[]` for performance considerations as to avoid transforming byte arrays back and forth to strings. However, such typing might come accross as complex and not user friendly.

Reply:
You are correct. Those are good points, and both options are bad in their own way. I don't have a final idea here, thus, I only keep it theoretical right now, proving that API is extensible enough to allow that in the future.
One way to do it would be some functional inheritance, where internal function is `Function<byte[], byte[]>`, but we provide adapters between that and `Function<String, String>` (or other types). This way users have ability to sacrifice performance for nicer UX (arguably for most cases it doesn't matter)

Reply:
You are correct. Those are good points, and both options are bad in their own way. I don't have a final idea here, thus, I only keep it theoretical right now, proving that API is extensible enough to allow that in the future.
One way to do it would be some functional inheritance, where internal function is `Function<byte[], byte[]>`, but we provide adapters between that and `Function<String, String>` (or other types). This way users have ability to sacrifice performance for nicer UX (arguably for most cases it doesn't matter)